### PR TITLE
feat: client auto updater

### PR DESCRIPTION
**Pull request by conde2**
## TODO

  - [x] Fix loading and saving settings 
  - [x] Fix locales not showing up when first open the client
  - [x] Music is not playing after the updater runs
  
  I had to move the g_resources.init and g_platform.init maybe this caused some bugs?
  @mehah i would appreciate some help fixing this bugs.

### CODE REVIEW

File: modules/updater/updater.lua [hunk 1]
```diff
@@ -165,35 +163,40 @@ local function updateFiles(data, keepCurrentFiles)
...
+
       if binary then
         g_resources.updateExecutable(binary)
       end
+
       if restart then
-        g_app.restart()
+        g_logger.info("restart")
+        --g_app.exit()
+        -- TODO Restart
```
Comment (poxis) [inline modules/updater/updater.lua]: I have something like this
```
void Application::restart()
{
#if not(defined(ANDROID))
    STARTUPINFO si;
    PROCESS_INFORMATION pi;
    ZeroMemory(&si, sizeof(si));
    si.cb = sizeof(si);
    ZeroMemory(&pi, sizeof(pi));

    // Create the child process.
    if (CreateProcess((LPCSTR)g_resources.getBinaryName().c_str(), // path to the executable
        nullptr, // command line arguments
        nullptr, // process handle not inheritable
        nullptr, // thread handle not inheritable
        FALSE, // handle inheritance flag
        0, // creation flags
        nullptr, // use parent's environment block
        nullptr, // use parent's starting directory
        &si, // pointer to STARTUPINFO structure
        &pi) // pointer to PROCESS_INFORMATION structure
    ) {
        // Close process and thread handles.
        CloseHandle(pi.hProcess);
        CloseHandle(pi.hThread);
        
    } else {
        g_logger.fatal("Updater restart error. Please restart application");
    }
    quick_exit();
#else
    exit();
#endif
}
```

Comment (conde2) [reply modules/updater/updater.lua]: I will try and let you know, I still need to figure out how to properly update the .exe file

Comment (yrpen) [reply modules/updater/updater.lua]: @conde2 Does this work also for linux/mac executables?

Comment (conde2) [reply modules/updater/updater.lua]: The updater or this code from poxis?
The updater should work for mac/linux not sure the if the executable will update, I use windows so not able to test in other systems.

You can try removing the FREE_VERSION and test
```cpp
void ResourceManager::updateExecutable(std::string fileName)
{
#if defined(ANDROID) || defined(FREE_VERSION)
    g_logger.fatal("Executable cannot be updated on android or in free version");
```

**Comment by poxis** [comment]
@conde2  try (client.lua) https://pastebin.com/9k69Kftp
so either change the loading order or just use 'scheduleEvent' to wait for everything to load and then open the login or language selection window 

the problem is probably here (the client is already run so it won't get this feature) 
https://github.com/mehah/otclient/blob/750077a017b2dc17eb6b89daa0711ea8f7b87962/modules/client/client.lua#L56


edit:
what the problem is with 'Fix loading and saving settings' ? 

edit2:
you can and even need to add to the updater options when there is no given module (folder) then the system creates it because at the moment there is an error such as : Cannot write file: modules/game_newfolder/newfolder.lua

modules/game_newfolder no exist in client 

my solution: https://pastebin.com/fAj6tShd

**Comment by conde2** [comment]
> @conde2 try (client.lua) https://pastebin.com/9k69Kftp so either change the loading order or just use 'scheduleEvent' to wait for everything to load and then open the login or language selection window
> 
> the problem is probably here (the client is already run so it won't get this feature)
> 
> https://github.com/mehah/otclient/blob/750077a017b2dc17eb6b89daa0711ea8f7b87962/modules/client/client.lua#L56
> 
> edit: what the problem is with 'Fix loading and saving settings' ?
> 
> edit2: you can and even need to add to the updater options when there is no given module (folder) then the system creates it because at the moment there is an error such as : Cannot write file: modules/game_newfolder/newfolder.lua
> 
> modules/game_newfolder no exist in client
> 
> my solution: https://pastebin.com/fAj6tShd


Nice catch!
Thanks for pointing me where the problem is, it will be much easier to fix it =)

**Comment by conde2** [comment]
@poxis 
I fixed the problems you pointed out

There was not problem with loading and saving the settings the problem was that I had a .otml file in 2 places and it loaded from the first and saved on the second one

**Pull request merged**